src/wash/wash.cpp

```diff
@@ -0,0 +1,132 @@
+        std::vector<Particle> particles;
+        std::unordered_map<std::string, double> variables;
+        std::string simulation_name;
+        std::string output_file_name;
+    }
+
+    ForceKernel::ForceKernel(const ForceFuncT func) : func(func) {}
+
+    void ForceKernel::exec() const {
+        for (auto& p : get_particles()) {
+            func(p, neighbors_kernel(p));
+        }
+    }
+
+    UpdateKernel::UpdateKernel(const UpdateFuncT func) : func(func) {}
+
+    void UpdateKernel::exec() const {
+        for (auto& p : get_particles()) {
+            func(p);
+        }
+    }
+
+    ReductionKernel::ReductionKernel(const MapFuncT map_func, const ReduceFuncT reduce_func, const double seed,
+                                     const std::string variable)
+        : map_func(map_func), reduce_func(reduce_func), seed(seed), variable(variable) {}
+
+    void ReductionKernel::exec() const {
+        auto result = seed;
+        for (auto& p : get_particles()) {
+            result = reduce_func(result, map_func(p));
+        }
+        set_variable(variable, result);
+    }
+
+    VoidKernel::VoidKernel(const VoidFuncT func) : func(func) {}
+
+    void VoidKernel::exec() const { func(); }
+
+    uint64_t get_max_iterations() { return max_iterations; }
+
+    void set_max_iterations(const uint64_t iterations) { max_iterations = iterations; }
+
+    void add_force_scalar(const std::string force) { forces_scalar.push_back(force); }
+
+    void add_force_vector(const std::string force) { forces_vector.push_back(force); }
+
+    void add_variable(const std::string variable, double init_value) { variables.emplace(variable, init_value); }
+
+    void add_init_kernel(const VoidFuncT func) { init_kernels.push_back(std::make_unique<VoidKernel>(func)); }
+
+    void add_force_kernel(const ForceFuncT func) { loop_kernels.push_back(std::make_unique<ForceKernel>(func)); }
+
+    void add_update_kernel(const UpdateFuncT func) { loop_kernels.push_back(std::make_unique<UpdateKernel>(func)); }
+
+    void add_reduction_kernel(const MapFuncT map_func, const ReduceFuncT reduce_func, const double seed,
+                    const std::string variable) {
+        loop_kernels.push_back(std::make_unique<ReductionKernel>(map_func, reduce_func, seed, variable));
+    }
+
+    void add_void_kernel(const VoidFuncT func) { loop_kernels.push_back(std::make_unique<VoidKernel>(func)); }
+
+    void set_neighbor_search_radius(const double radius) {
+        neighbors_kernel = [radius](const Particle& p) { return get_neighbors(p, radius); };
+    }
+
+    void set_neighbor_search_kernel(const NeighborsFuncT func) { neighbors_kernel = func; }
+
+    Particle& create_particle(const double density, const double mass, const double smoothing_length,
+                              const SimulationVecT pos, const SimulationVecT vel, const SimulationVecT acc) {
+        auto id = particles.size();
+        auto& p = particles.emplace_back(id);
+        p.set_density(density);
```

Comment:
We could add another constructor for particles that takes in all the initialisation params, so we don't need all these individual lines and just have `p.emplace_back(id, density, mass, smoothing_length, pos, vel, acc`) instead

Reply:
Done